### ISSUE TITLE
Incorrect state for pods with crashlooping containers

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#15671]
With the enhancement of container restart backoff, pod phase may not progress monotonically.

A pod is considered running if at least one container in the pod has been started. If the container happens to be crashlooping, the restart backoff would eventually kick in. From that point on, the container status becomes waiting, and [kubelet determines the phase](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2497) as **pending**.

Once the pod transitions from `Running` to `Pending`, it's impossible to retrieve logs from the crashlooping container using kubectl. This prevents the users from debugging the cause of container failures.
```console
$ kubectl logs -p heapster-v10-3ayyj --namespace=kube-system
Error from server: Internal error occurred: Pod "heapster-v10-3ayyj" in namespace "kube-system" : pod is not in 'Running', 'Succeeded' or 'Failed' state - State: "Pending"
```

An easy fix is to check `LastTerminationState` of a container before determining the pod phase.

/cc @samsabed

**Comment by samsabed** [comment]
@yujuhong so change it to stay in running when one of the containers is not running ?

**Comment by yujuhong** [comment]
> @yujuhong so change it to stay in running when one of the containers is not running ?

Hm...once the pod has transitioned to "running", it should stay as "running" or transition to a final phase based on the restart policy. It should never go back to "pending" though.
This wasn't a problem before we had restart backoff, but now there could be a time when all containers in a pod are in the "waiting" status, and none of them are running.

I think we should treat the pod as running as long as every container in the pod matches one of the following conditions (@bgrant0607, correct me if I am wrong):
 1. container is running.
 2. container is terminated.
 3. container is waiting (due to backoff) but container has been run before (non-nil `LastTerminationState`)

If *all* containers are terminated, we may transition the pod into phase Succeeded or Failed based on the termination status and the restart policy.

One exception I can think of is the "failed to start" containers, which appear as terminated containers. @bgrant0607, if kubelet has tried starting all containers in the pod, but all of the failed to run. Can we consider the pod running?

**Comment by bgrant0607** [comment]
@yujuhong is correct.

**Comment by bgrant0607** [comment]
Rather than phase, ready, at both the container level and pod level, should be used to infer whether the containers are capable of doing anything useful.

**Comment by samsabed** [comment]
probably best to do a quick phase fix and look into conditions (BTW in this instance the container is terminated and we want to access the logs)

for the conditions part, do you think a pod condition  ready=True should take into account readiness / liveness
@bgrant0607 @yujuhong

**Comment by yujuhong** [comment]
> for the conditions part, do you think a pod condition ready=True should take into account readiness / liveness

AFAIK, pod ready condition already takes into account the readiness of the containers.

Agreed that we should fix this sooner than later.

**Comment by dchen1107** [comment]
Ahh, I ran into this issue yesterday for testing. Yes, Pod Phase shouldn't switch back to Pending once all containers in such pod run once. This is a regression, and we should fix for 1.1. I will send pr shortly.

**Comment by yujuhong** [comment]
@dchen1107, please cherry-pick #15799, thanks!
